src/providers/azurestack/mod.rs

```diff
@@ -0,0 +1,369 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! AzureStack provider, metadata and wireserver fetcher.
+
+mod crypto;
+mod goalstate;
+
+use std::collections::HashMap;
+use std::net::IpAddr;
+
+use anyhow::{anyhow, bail, Context, Result};
+use openssh_keys::PublicKey;
+use reqwest::header::{HeaderName, HeaderValue};
+use serde_derive::Deserialize;
+use slog_scope::warn;
+
+use self::crypto::x509;
+use crate::providers::MetadataProvider;
+use crate::retry;
+use nix::unistd::Uid;
+
+#[cfg(test)]
+mod mock_tests;
+
+static HDR_AGENT_NAME: &str = "x-ms-agent-name";
+static HDR_VERSION: &str = "x-ms-version";
+static HDR_CIPHER_NAME: &str = "x-ms-cipher-name";
+static HDR_CERT: &str = "x-ms-guest-agent-public-x509-cert";
+
+const MS_AGENT_NAME: &str = "com.coreos.afterburn";
+const MS_VERSION: &str = "2012-11-30";
+const SMIME_HEADER: &str = "\
+MIME-Version:1.0
+Content-Disposition: attachment; filename=/home/core/encrypted-ssh-cert.pem
+Content-Type: application/x-pkcs7-mime; name=/home/core/encrypted-ssh-cert.pem
+Content-Transfer-Encoding: base64
+
+";
+
+/// This is a known working wireserver endpoint within AzureStack.
+/// See: https://blogs.msdn.microsoft.com/mast/2015/05/18/what-is-the-ip-address-168-63-129-16/
+#[cfg(not(test))]
+const FALLBACK_WIRESERVER_ADDR: [u8; 4] = [168, 63, 129, 16]; // for grep: 168.63.129.16
+
+macro_rules! ready_state {
+    ($container:expr, $instance:expr, $incarnation:expr) => {
+        format!(r#"<?xml version="1.0" encoding="utf-8"?>
+<Health xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
+  <GoalStateIncarnation>{}</GoalStateIncarnation>
+  <Container>
+    <ContainerId>{}</ContainerId>
+    <RoleInstanceList>
+      <Role>
+        <InstanceId>{}</InstanceId>
+        <Health>
+          <State>Ready</State>
+        </Health>
+      </Role>
+    </RoleInstanceList>
+  </Container>
+</Health>
+"#,
+                $incarnation, $container, $instance)
+    }
+}
+
+#[derive(Debug, Deserialize, Clone)]
+struct Versions {
+    #[serde(rename = "Supported")]
+    pub supported: Supported,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+struct Supported {
+    #[serde(rename = "Version", default)]
+    pub versions: Vec<String>,
+}
+
+#[derive(Debug, Clone)]
+pub struct AzureStack {
+    client: retry::Client,
+    endpoint: IpAddr,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+struct InstanceMetadata {
+    #[serde(rename = "vmName", default)]
+    pub vm_name: String,
+    #[serde(rename = "subscriptionId", default)]
+    pub subscription_id: String,
+}
```

Comment:
Should be able to simplify to

```suggestion
#[derive(Debug, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
struct InstanceMetadata {
    pub vm_name: String,
    pub subscription_id: String,
}
```

Do we actually need the `default` here? I.e. do we always expect those fields to exist, or do we want to support them not existing? In case of the former, we should just drop `default`. In case of the latter, it would be cleaner to use an `Option<String>` instead.